File: src/nix_spotlight/dock.py
"""Dock syncing via dockutil."""

import shutil
import subprocess
from pathlib import Path

from .types import DockSyncResult


def _iter_nix_dock_items(output: str) -> list[tuple[str, str]]:
    """Extract dock item name/path pairs that reference /nix/store."""
    items: list[tuple[str, str]] = []
    for line in output.splitlines():
        if not line.strip():
            continue
        if "/nix/store" not in line:
            continue

        parts = line.split("\t", maxsplit=1)
        name = parts[0].strip()
        path = parts[1].strip() if len(parts) > 1 else ""
        items.append((name, path))

    return items


def sync_dock(apps: list[Path], dockutil_path: str | None = None) -> DockSyncResult:
    """Update dock persistent items pointing to /nix/store.

    Finds pinned dock items with /nix/store paths and updates them
    to point to the new trampoline locations.

    Args:
        apps: List of trampoline app paths
        dockutil_path: Path to dockutil binary (auto-detected if None)

    Returns:
        DockSyncResult with counts of updated, skipped items and any errors

    """
    dockutil = dockutil_path or shutil.which("dockutil")
    if not dockutil:
        return DockSyncResult()

    result = subprocess.run(
        [dockutil, "-L"],
        capture_output=True,
        text=True,
        check=False,
    )

    if result.returncode != 0:
        return DockSyncResult(errors=(f"dockutil -L failed: {result.stderr}",))

    app_stems = {app.stem: app for app in apps}
    updated = 0
    skipped = 0
    errors: list[str] = []

    for name, path in _iter_nix_dock_items(result.stdout):
        path_stem = Path(path).stem if path else ""

        trampoline = app_stems.get(name) or app_stems.get(path_stem)
        if not trampoline:
            skipped += 1
            continue

        add_result = subprocess.run(
            [dockutil, "--add", str(trampoline.resolve()), "--replacing", name],
            capture_output=True,
            text=True,
            check=False,
        )

        if add_result.returncode != 0:
            errors.append(f"Failed to update {name}: {add_result.stderr}")
        else:
            updated += 1

    return DockSyncResult(updated=updated, skipped=skipped, errors=tuple(errors))

File: dock.go
package nixspotlight

import (
	"os/exec"
	"path/filepath"
	"strings"
)

// DockSyncResult is the result of a dock sync operation.
type DockSyncResult struct {
	Updated int
	Skipped int
	Errors  []string
}

// SyncDock updates dock items pointing to /nix/store paths to use trampoline locations.
func SyncDock(apps []string, dockutilPath string) DockSyncResult {
	dockutil := dockutilPath
	if dockutil == "" {
		var err error
		dockutil, err = exec.LookPath("dockutil")
		if err != nil || dockutil == "" {
			return DockSyncResult{}
		}
	}

	stdout, stderr, ok := runDockutil(dockutil, "-L")
	if !ok {
		return DockSyncResult{Errors: []string{"dockutil -L failed: " + stderr}}
	}

	appStems := make(map[string]string, len(apps))
	for _, app := range apps {
		appStems[strings.TrimSuffix(filepath.Base(app), ".app")] = app
	}

	result := DockSyncResult{}
	for _, line := range strings.Split(stdout, "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		if !strings.Contains(line, "/nix/store") {
			continue
		}

		parts := strings.SplitN(line, "\t", 2)
		name := strings.TrimSpace(parts[0])
		path := ""
		if len(parts) > 1 {
			path = strings.TrimSpace(parts[1])
		}

		trampoline := appStems[name]
		if trampoline == "" && path != "" {
			trampoline = appStems[strings.TrimSuffix(filepath.Base(path), filepath.Ext(filepath.Base(path)))]
		}
		if trampoline == "" {
			result.Skipped++
			continue
		}

		resolved := trampoline
		if resolvedPath, err := filepath.EvalSymlinks(trampoline); err == nil {
			resolved = resolvedPath
		}

		_, addStderr, addOK := runDockutil(dockutil, "--add", resolved, "--replacing", name)
		if !addOK {
			result.Errors = append(result.Errors, "Failed to update "+name+": "+addStderr)
			continue
		}
		result.Updated++
	}

	return result
}

func runDockutil(dockutil string, args ...string) (stdout string, stderr string, ok bool) {
	output, err := exec.Command(dockutil, args...).Output() //nolint:noctx // short-lived CLI, no context needed
	if err != nil {
		if exitErr, isExitErr := err.(*exec.ExitError); isExitErr {
			return string(output), string(exitErr.Stderr), false
		}
		return string(output), "", false
	}
	return string(output), "", true
}
